File: 2794-maximum-number-of-moves-in-a-grid/2794-maximum-number-of-moves-in-a-grid.java
class Solution {
    public int maxMoves(int[][] grid) {
        int row=grid.length;
        int col=grid[0].length;
        int[][] tabulation=new int[row][col];
        for(int i=0;i<row;i++){
            for(int j=0;j<col;j++){
                tabulation[i][j]=-1;
            }
            tabulation[i][0]=0;
        }
        int[][] directions={{-1,-1},{0,-1},{1,-1}};
        for(int j=1;j<col;j++){
            for(int i=0;i<row;i++){
                for(int[] d:directions){
                    int nx=i+d[0];
                    int ny=j+d[1];
                    if(nx<0 || ny<0 || nx>=row || ny>=col )continue;
                    if(grid[nx][ny]<grid[i][j] && tabulation[nx][ny] != -1){
                        tabulation[i][j]=Math.max(tabulation[nx][ny]+1,tabulation[i][j]);
                    }
                }
            }
        }
        int maxi=0;
        for(int i=0;i<row;i++){
            for(int j=0;j<col;j++){
                maxi = Math.max(tabulation[i][j], maxi);
            }
        }
        return maxi;
    }
}

File: 2794-maximum-number-of-moves-in-a-grid/2794-maximum-number-of-moves-in-a-grid.py
class Solution:
    def maxMoves(self, grid: List[List[int]]) -> int:
        def display(mat):
            for row in mat:
                print(row)
            print('-------------------------')
        row,col=len(grid),len(grid[0])
        tabulation=[[-1]*col for _ in range(row)]
        for i in range(row):
            tabulation[i][0]=0
        directions=[(-1,-1),(0,-1),(1,-1)]
        for j in range(1,col):
            for i in range(row):
                for dx,dy in directions:
                    nx,ny=i+dx,j+dy
                    if nx<0 or ny<0 or nx>=row or ny>=col:
                        continue
                    elif grid[nx][ny]<grid[i][j]:
                        if tabulation[nx][ny]!=-1:
                            tabulation[i][j]=max(tabulation[nx][ny]+1,tabulation[i][j])
        display(tabulation)
        display(grid)
        maxi=0
        for i in range(row):
            for j in range(col):
                maxi=max(tabulation[i][j],maxi)
        return maxi
